File: frontend-v2/src/api/notes.ts
import { apiDelete, apiGet, apiPatch, apiPost } from './client';

export interface Note {
  id: string;
  user_id: number;
  title: string | null;
  content: string;
  created_at: string;
  updated_at: string;
}

export async function listNotes(): Promise<Note[]> {
  return apiGet<Note[]>('/notes');
}

export async function createNote(body: { title?: string | null; content: string }): Promise<Note> {
  return apiPost<Note>('/notes', body);
}

export async function updateNote(
  id: string,
  patch: { title?: string | null; content?: string },
): Promise<Note> {
  return apiPatch<Note>(`/notes/${id}`, patch);
}

export async function deleteNote(id: string): Promise<void> {
  return apiDelete(`/notes/${id}`);
}

File: backend/cara/services/notes.py
"""Note persistence helpers."""

from __future__ import annotations

import uuid
from datetime import datetime, timezone

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from cara.models import Note


async def list_notes(session: AsyncSession, *, user_id: int) -> list[Note]:
    stmt = (
        select(Note).where(Note.user_id == user_id).order_by(Note.updated_at.desc())
    )
    return list((await session.execute(stmt)).scalars().all())


async def create_note(
    session: AsyncSession, *, user_id: int, title: str, body: str = ""
) -> Note:
    note = Note(user_id=user_id, title=title or "Nota senza titolo", body=body)
    session.add(note)
    await session.flush()
    return note


async def get_note(
    session: AsyncSession, note_id: uuid.UUID, *, user_id: int
) -> Note | None:
    stmt = select(Note).where(Note.id == note_id, Note.user_id == user_id)
    return (await session.execute(stmt)).scalar_one_or_none()


async def update_note(
    session: AsyncSession,
    note_id: uuid.UUID,
    *,
    user_id: int,
    title: str | None = None,
    body: str | None = None,
) -> Note | None:
    note = await get_note(session, note_id, user_id=user_id)
    if note is None:
        return None
    if title is not None:
        note.title = title or "Nota senza titolo"
    if body is not None:
        note.body = body
    # Esplicito (no `onupdate=func.now()` sul modello — vedi commento nel
    # modello: MissingGreenlet su async se la colonna va refresh-required).
    note.updated_at = datetime.now(timezone.utc)
    await session.flush()
    return note


async def delete_note(
    session: AsyncSession, note_id: uuid.UUID, *, user_id: int
) -> bool:
    note = await get_note(session, note_id, user_id=user_id)
    if note is None:
        return False
    await session.delete(note)
    return True
